File: day2/go/main.go
package main

import (
	"fmt"
	"io/ioutil"
	"strconv"
	"strings"
)

func main() {
	instructions, err := ioutil.ReadFile("input.test")
	if err != nil {
		fmt.Println("Error reading file", err)
		return
	}
	part1(string(instructions))
	part2(string(instructions))
}

func part1(instructions string) {
	curx, curd := 0, 0
	for _, instruction := range strings.Split(string(instructions), "\r\n") {
		direction, value := parseInstruction(instruction)
		switch direction {
		case "forward":
			curx += value
		case "down":
			curd += value
		case "up":
			curd -= value
		}
	}
	fmt.Printf("Day 2 - Part 1: %v\n", curx*curd)
}

func part2(instructions string) {
	curx, curd, cura := 0, 0, 0
	for _, instruction := range strings.Split(string(instructions), "\r\n") {
		direction, value := parseInstruction(instruction)
		switch direction {
		case "forward":
			curx += value
			curd += cura * value
		case "down":
			cura += value
		case "up":
			cura -= value
		}
	}
	fmt.Printf("Day 2 - Part 2: %v\n", curx*curd)
}

func parseInstruction(instruction string) (string, int) {
	is := strings.Split(instruction, " ")
	amount, _ := strconv.Atoi(is[1])
	return is[0], amount
}

File: day2/rust/main.rs
use std::fs;

fn main() {
    let instruction_list = fs::read_to_string("../input")
        .expect("Something went wrong reading the file");

    let instructions: Vec<&str> = instruction_list.split("\r\n").collect();
    part_1(&instructions);
    part_2(&instructions);
}

fn part_1(instructions: &Vec<&str>) {
    let mut current_x = 0;
    let mut current_d = 0;
    for instruction in instructions {
        let (direction, value) = parse_instruction(instruction);
        match direction {
            "forward" => current_x += value,
            "down" => current_d += value,
            "up" => current_d -= value,
            _ => println!("Command Not Supported")
        }
    }
    println!("{:?}", current_d * current_x);
}

fn part_2(instructions: &Vec<&str>) {
    let mut current_x = 0;
    let mut current_d = 0;
    let mut current_aim = 0;
    for instruction in instructions {
        let (direction, value) = parse_instruction(instruction);
        match direction {
            "forward" => {
                current_x += value;
                current_d += current_aim * value
            },
            "down" => current_aim += value,
            "up" => current_aim -= value,
            _ => println!("Command Not Supported")
        }
    }
    println!("{:?}", current_d * current_x);
}

fn parse_instruction(instruction: &str) -> (&str, i32) {
    let pi: Vec<&str> = instruction.split(" ").collect();
    return (pi[0], pi[1].parse::<i32>().unwrap());
}